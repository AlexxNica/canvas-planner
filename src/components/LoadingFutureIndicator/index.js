import React, { Component } from 'react';
import {bool, func} from 'prop-types';
import Button from 'instructure-ui/lib/components/Button';
import Container from 'instructure-ui/lib/components/Container';
import Spinner from 'instructure-ui/lib/components/Spinner';
import Typography from 'instructure-ui/lib/components/Typography';
import Waypoint from 'react-waypoint';
import formatMessage from 'format-message';

export default class LoadingFutureIndicator extends Component {
  static propTypes = {
    loadingFuture: bool,
    allFutureItemsLoaded: bool,
    onLoadMore: func,
  }

  static defaultProps: {
    loadingFuture: false,
    allFutureItemsLoaded: false,
    onLoadMore: () => {},
  }

  handleLoadMoreButton = () => {
    this.props.onLoadMore({setFocusAfterLoad: true});
  }

  handleWaypoint = () => {
    this.props.onLoadMore();
  }

  renderLoadMore () {
    if (!this.props.loadingFuture && !this.props.allFutureItemsLoaded) {
      return <Button variant="link" onClick={this.handleLoadMoreButton}>
        {formatMessage('Load more')}
      </Button>;
    }
  }

  renderLoading () {
    if (this.props.loadingFuture && !this.props.allFutureItemsLoaded) {
      return <Container>
        <Container display="inline">
          <Spinner size="small" margin="0 x-small 0 0" title={formatMessage('Loading...')} />
        </Container>
        <Typography size="small" color="secondary">
          {formatMessage('Loading...')}</Typography>
      </Container>;
    }
  }

  renderEverythingLoaded () {
    if (this.props.allFutureItemsLoaded) {
      return <Typography color="secondary" size="small">
        {formatMessage('All items loaded')}</Typography>;
    }
  }

  renderWaypoint () {
    if (!this.props.loadingFuture && !this.props.allFutureItemsLoaded){
      return <Waypoint onEnter={this.handleWaypoint} />;
    }
  }

  render () {
    return <div>
      {this.renderWaypoint()}
      <Container as="div" padding="x-large" textAlign="center">
        {this.renderLoadMore()}
        {this.renderLoading()}
        {this.renderEverythingLoaded()}
      </Container>
    </div>;
  }
}
