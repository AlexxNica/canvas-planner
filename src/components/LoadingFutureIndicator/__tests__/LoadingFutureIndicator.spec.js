import React from 'react';
import {shallow} from 'enzyme';
import LoadingFutureIndicator from '../index';

it('renders load more by default', () => {
  const wrapper = shallow(<LoadingFutureIndicator />);
  expect(wrapper).toMatchSnapshot();
});

it('renders loading when indicated', () => {
  const wrapper = shallow(<LoadingFutureIndicator loadingFuture />);
  expect(wrapper).toMatchSnapshot();
});

it('renders all future items loaded regardless of other props', () => {
  const wrapper = shallow(<LoadingFutureIndicator loadingFuture allFutureItemsLoaded />);
  expect(wrapper).toMatchSnapshot();
});

it('invokes the callback when the waypoint is triggered', () => {
  const mockLoad = jest.fn();
  const wrapper = shallow(<LoadingFutureIndicator onLoadMore={mockLoad} />);
  wrapper.instance().handleWaypoint();
  expect(mockLoad).toHaveBeenCalledWith();
});

it('invokes the callback when loading more button is clicked', () => {
  const mockLoad = jest.fn();
  const wrapper = shallow(<LoadingFutureIndicator onLoadMore={mockLoad} />);
  wrapper.find('Button').simulate('click');
  expect(mockLoad).toHaveBeenCalledWith({setFocusAfterLoad: true});
});
