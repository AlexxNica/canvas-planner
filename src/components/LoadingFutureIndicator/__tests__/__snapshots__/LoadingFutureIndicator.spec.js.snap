// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders all future items loaded regardless of other props 1`] = `
<div>
  <Container
    as="div"
    display={null}
    padding="x-large"
    textAlign="center"
  >
    <Typography
      as="span"
      color="secondary"
      letterSpacing="normal"
      size="small"
    >
      All items loaded
    </Typography>
  </Container>
</div>
`;

exports[`renders load more by default 1`] = `
<div>
  <Container
    as="div"
    display={null}
    padding="x-large"
    textAlign="center"
  >
    <Button
      as="button"
      buttonRef={[Function]}
      fluidWidth={false}
      margin="0"
      onClick={[Function]}
      size="medium"
      type="button"
      variant="link"
    >
      Load more
    </Button>
  </Container>
  <Waypoint
    bottomOffset="0px"
    fireOnRapidScroll={true}
    horizontal={false}
    onEnter={[Function]}
    onLeave={[Function]}
    onPositionChange={[Function]}
    topOffset="0px"
  />
</div>
`;

exports[`renders loading when indicated 1`] = `
<div>
  <Container
    as="div"
    display={null}
    padding="x-large"
    textAlign="center"
  >
    <Container
      display={null}
    >
      <Container
        display="inline"
      >
        <Spinner
          margin="0 x-small 0 0"
          size="small"
          title="Loading..."
          variant="default"
        />
      </Container>
      <Typography
        as="span"
        color="secondary"
        letterSpacing="normal"
        size="small"
      >
        Loading...
      </Typography>
    </Container>
  </Container>
</div>
`;
